Implement Skeleton Editor Gizmo

TokageItLab opened the pull request:
Skeleton's TreeEditor was implemented in #36409, but the gizmo for editing was not (ref: [584](https://github.com/godotengine/godot-proposals/issues/584) and [#1767](https://github.com/godotengine/godot-proposals/issues/1767)). So, the gizmo in `SkeletonEditor`, which was developed in the 3.2.x branch, is now complete tentatively. This is a draft PR for porting to 4.0.

https://www.youtube.com/watch?v=RjJiEo8uqPw

Dev branch for 3.2.x is [here](https://github.com/TokageItLab/godot/tree/pose-edit-mode).

## Features

![pr1](https://user-images.githubusercontent.com/61938263/106864740-a119f200-670d-11eb-949e-035335f65b48.png)

When you select skeleton and bone edit tool, marker circles will appear on the joints. If you select a marker, a transform gizmo will appear, ~which looks the same as `SpatialTransformGizmo` but behaves a bit differently. Basically, in character animation, there is no editing other than rotate. So to make the editing process smoother, you can make selections in all `Move/Rotate/Scale` modes. Instead, there is no free transform mode, which I may implement somewhere else.~ the behavior was unified by SubGizmo implementation.

![pr2](https://user-images.githubusercontent.com/61938263/106864763-a6773c80-670d-11eb-8030-edd671b8c634.gif)

There is a `PoseMode` and a `RestMode`; during `RestMode`, `Pose` and `CustomPose` are temporarily disabled, and `RestMode` will be canceled when you exit the `SkeletonEditor`. However, there is no `CustomPoseMode`, because there is no way to disable only the `Pose` while the `CustomPose` is enabled.

![pr3](https://user-images.githubusercontent.com/61938263/106864775-a9722d00-670d-11eb-8b28-a71f596c5db6.png)

Add the following options to `SkeletonButton`'s popups in toolber

- Init pose
- Insert key of all bone poses
- Insert key of bone poses already exist track
- Apply current pose to rest

All of these are basically applied to all bones. *Init pose* initializes the Pose, but not the `CustomPose`. *Insert key* is activated when the animation player has a track, similar to `Spatial`'s `KeyButton`. Be careful with *Apply current pose to rest*. It initializes the current `Pose` and `CustomPose`, and changes `Rest`.

## Changes/Bugfixes

### SkeletonEditor
- Added `int selected_bone`
- Removed the *Pose Enabled* checkbox in `SkeletonBoneEditor` to prevent conflicts in `RestMode`

### SpatialEditorGizmo
- Skeleton's gizmo is now at the front of the screen when skeleton is selected
- Show the local axis of the bones when selecting a skeleton
- Highlight the currently selected bone
- Added *selected bone color* and *bone axis length* to `EditorSettings`

### SpatialEditorPlugin
- ~Added `Vector<Transform> external`~
- ~Added `ToolExternal` to `ToolMode`~
- ~`forward_spatial_gui_input()` passes the id of the Viewport~
- ~Some private functions changed to public~

These were superseded SubGizmo implementation.

### AnimationTrackEditor
- Added `has_transform_track()`

## Done

- [x] Porting for 4.0
- [x] Fix depth drawing order in gizmo (with the shader by @lyuma )
- [x] Consider the size and shape of the joint marker
- [x] Adopt SubGizmos
- [x] Behavior for negative scales is not ensured
- [x] Set readonly to bone pose in rest mode after marge #51722
- [x] Octahedron as bone gizmo in `SpatialEditorGizmo` weight assignment and updated timing are strange (superseded selectable gizmo shape with Wire shape)

## Other Issues (Todo in the future)
- ~~Need to fix some function (inserting multiple tracks has been broken) #51459~~ Fixed
- `SkeletonEditor` does not support multiple selection
- Need independent bone handling (with changing skin pose) in edit rest option for adjust initial rest
- ~~#52627~~ Fixed

TokageItLab committed:
Implemented SkeletonEditorGizmo

Co-authored-by: Lyuma <xn.lyuma@gmail.com>

JFonS commented:
I think you missed most of the changes in your commit, I can only see a modification in `skeleton_3d_editor_plugin.cpp` :)

fire commented:
Can you write a description of the ExternalToolMode system? This is the api addition to Node3d gizmos we wanted to add, but it's hard to describe to other contributors.

TokageItLab commented:
@JFonS @fire Porting is done. There are a few issues that need to be fixed, but you can check that it works for now.

- The depth drawing order in gizmo is messed up. In 3.2.x I used a hack with vertex shader to change the draw order, but this is no longer available in 4.0, so I need to find another way.
- It was decided in discussion with reduz and others to implement SubGizmos first instead of ExternalMode in Node3DEditor.

TokageItLab commented:
Depth problem was solved. It's a bit hacky, but it maintains the drawing order in 3D space.

qarmin commented on editor/plugins/node_3d_editor_plugin.cpp:
```diff
@@ -6739,6 +6780,10 @@ Node3DEditor::Node3DEditor(EditorNode *p_editor) {
 	tool_button[TOOL_MODE_SCALE]->set_shortcut(ED_SHORTCUT("spatial_editor/tool_scale", TTR("Scale Mode"), KEY_R));
 	tool_button[TOOL_MODE_SCALE]->set_shortcut_context(this);
 
+	tool_button[TOOL_MODE_EXTERNAL] = memnew(Button);
```
This button isn't in scene tree, so it must be added to it with `add_child` or just deleted in destructor.

TokageItLab replied:
Ah! I was looking for what was causing the leak. I don't want to add that button to the scene, so I made it free in the destructor. Thank you very much!

fire commented:
According to reduz and jfons' review 4 months ago this is in limbo for a better design, but we can keep the patch maintained.

Note that the nullptr commit should be combined with the first commit.

fire commented:
Comments:

Move from red to dark orange.

    EDITOR_DEF("editors/3d_gizmos/gizmo_colors/selected_bone", Color(0.8f, 0.3f, 0.0f));

There was a bug where the selected bone was off by one.

https://github.com/godot-extended-libraries/godot-fire/commit/ef6fb114b5c91ca8516663f2ca01d12edaa91ccd#diff-2e3f5a3cec90a096257b4051e2b39ae52e208ab5d9331fc226f8f5e984b0f6e0R1594-R1692

TokageItLab commented:
@fire Sure, I just squashed that commit because I just wanted to see if it passed the CI test. Can I find any progress on SubGizmo, which we still discussed with reduz and jfons? I don't think there will be any progress on that... :'-(

TokageItLab commented:
Changed colour, fixed shadowcast bug, indexing of selected bones is correct.

JFonS commented:
I did an overall review and changes look mostly good, but there is too much code duplication between `Node3DEditor` and `Skeleton3DEditor`.

I plan to work on improvements to the gizmo system in the following days, so I will try to provide a better way to handle transforms in gizmos.

TokageItLab commented:
#50748 has been merged to implement SubGizmo. So I'm working on adapting it to SubGizmo.

TokageItLab commented:
I verified that it is possible to implement `Skeleton3DGizmo` with the current `SubGizmo` implementation. I'm still working on it to make it work without any problems.

https://user-images.githubusercontent.com/61938263/126917543-40698667-7b7d-4727-ae39-68b773ef045f.mov

JFonS commented:
Here's the code for `MyNode` I used in the example, as I mentioned it just stores a vector of `Transform3D`:

<details>
<summary> MyNode source </summary>

```cpp
class MyNode : public Node3D {
	GDCLASS(MyNode, Node3D);

	Vector<Transform3D> xforms;

	public:

	void set_xforms(const Vector<Transform3D> &p_xforms);
	Vector<Transform3D> get_xforms() const;

	void _set_xforms(const Array &p_xforms);
	Array _get_xforms() const;

protected:

	static void _bind_methods();
};


////////////////////////////////////////////


void MyNode::set_xforms(const Vector<Transform3D> &p_xforms) {
	if (p_xforms.size() != xforms.size()) {
		clear_subgizmo_selection();
	}
	xforms = p_xforms;
	update_gizmos();
}

Vector<Transform3D> MyNode::get_xforms() const {
	return xforms;
}

void MyNode::_set_xforms(const Array &p_xforms) {
	Vector<Transform3D> x;
	x.resize(p_xforms.size());
	for (int i = 0 ; i < x.size(); i++) {
		x.write[i] = p_xforms[i];
	}
	set_xforms(x);
}

Array MyNode::_get_xforms() const {
	Array a;
	a.resize(xforms.size());
	for (int i = 0; i < xforms.size(); i++) {
		a[i] = xforms[i];
	}
	return a;
}

void MyNode::_bind_methods() {
	ClassDB::bind_method(D_METHOD("set_xforms", "xforms"), &MyNode::_set_xforms);
	ClassDB::bind_method(D_METHOD("get_xforms"), &MyNode::_get_xforms);
	ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "xforms"), "set_xforms", "get_xforms");
}
```
</details>

I opened #50933 with some fixes to the subgizmo system, mainly the scaled node issue and the input priority of subgizmo selection.

I did a quick review of this PR so far, and I noticed you are using global transforms on the subgizmo methods. The whole system was designed and tested to use local transforms, although it may be possible to use global transforms, I would advise against it. I added some more clarification on the documentation to make this clearer.

fire commented:
I tested the pr.

Noticed that there's no undo redo.

1. click on point
2. move rotation
3. move rotation
4. press control z to undo
5. nothing happens

fire commented:
I am unsure if this crash is related.

```
EditorInspectorPluginSkeleton::forward_spatial_gui_input(Camera3D*, Ref<InputEvent> const&) (Unknown Source:0)
EditorPluginList::forward_spatial_gui_input(Camera3D*, Ref<InputEvent> const&, bool) (Unknown Source:0)
Node3DEditorViewport::_sinput(Ref<InputEvent> const&) (Unknown Source:0)
void call_with_variant_args_helper<Node3DEditorViewport, Ref<InputEvent> const&, 0ul>(Node3DEditorViewport*, void (Node3DEditorViewport::*)(Ref<InputEvent> const&), Variant const**, Callable::CallError&, IndexSequence<0ul>) (Unknown Source:0)
Object::emit_signal(StringName const&, Variant const**, int) (Unknown Source:0)
Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (Unknown Source:0)
Viewport::_gui_call_input(Control*, Ref<InputEvent> const&) (Unknown Source:0)
Viewport::_gui_input_event(Ref<InputEvent>) (Unknown Source:0)
Viewport::input(Ref<InputEvent> const&, bool) (Unknown Source:0)
Window::_window_input(Ref<InputEvent> const&) (Unknown Source:0)
void call_with_variant_args_helper<Window, Ref<InputEvent> const&, 0ul>(Window*, void (Window::*)(Ref<InputEvent> const&), Variant const**, Callable::CallError&, IndexSequence<0ul>) (Unknown Source:0)
DisplayServerX11::_dispatch_input_event(Ref<InputEvent> const&) (Unknown Source:0)
Input::_parse_input_event_impl(Ref<InputEvent> const&, bool) (Unknown Source:0)
Input::flush_accumulated_events() (Unknown Source:0)
DisplayServerX11::process_events() (Unknown Source:0)
OS_LinuxBSD::run() (Unknown Source:0)
main (Unknown Source:0)
```

TokageItLab commented:
@fire It's still a work in progress, we may have misapplied the Scale in Node3DGizmo in the first place, so I'm working on fixing that.

TokageItLab commented:
I think the porting is almost done, except for problem #51415. There are some problems as described at the bottom of the overview, but for now I want to determine the behavior of the basics. As I mentioned at the bottom of the overview, there are a few problems, but I would like to determine the basic behavior for now.

TokageItLab commented:
@JFonS I made a few changes to the Gizmo.
- Implemented `set_gizmo_visible()` in `Node3DEditor` to hide gyro gizmo for transform
- Implemented `_set_subgizmo_selection()` in `Node3DEditor` to allow subgizmo selection without `intersect_ray()`; it will be called by Node just like `clear_subgizmo_selection()

TokageItLab commented:
> I think the porting is almost done, except for problem #51415. And there are some issues as noted at the bottom of the overview, but for now I want to determine the behavior of the basics.

Fixed by #51459.
When generating bone animation, we need to make this change because we need to insert track and keys of multiple bones at once.

JFonS commented:
Nice work :) I will review it soon-ish whenever I have some free time.

JFonS commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -666,22 +804,255 @@ void Skeleton3DEditor::_bind_methods() {
+		Vector<Ref<Node3DGizmo>> gizmos = skeleton->get_gizmos();
+		for (int i = 0; i < gizmos.size(); i++) {
+			Ref<EditorNode3DGizmo> seg = gizmos[i];
+			if (seg->get_plugin()->get_gizmo_name() == "Skeleton3D") {
```
This check can be changed to:
```
Ref<Skeleton3DGizmoPlugin> plugin = gizmos[i]->get_plugin();
if (plugin.is_valid()) {
    ...
}
```

TokageItLab replied:
finally
```cpp
Vector<Ref<Node3DGizmo>> gizmos = skeleton->get_gizmos();
for (int i = 0; i < gizmos.size(); i++) {
	Ref<EditorNode3DGizmo> gizmo = gizmos[i];
	if (!gizmo.is_valid()) {
		continue;
	}
	Ref<Skeleton3DGizmoPlugin> plugin = gizmo->get_plugin();
	if (!plugin.is_valid()) {
		continue;
	}
	if (skeleton->get_selected_bone() != -1) {
		skeleton->set_subgizmo_selection(gizmo, 0, skeleton->get_bone_global_pose_no_override(skeleton->get_selected_bone()));
	} else {
		skeleton->set_subgizmo_selection(gizmo, 0);
	}
	break;
}
```

JFonS commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -692,16 +1063,374 @@ void EditorInspectorPluginSkeleton::parse_begin(Object *p_object) {
+	selected_mat = Ref<ShaderMaterial>(memnew(ShaderMaterial));
+	selected_sh = Ref<Shader>(memnew(Shader));
+	selected_sh->set_code(" \
+		shader_type spatial; \
+		render_mode unshaded, shadows_disabled; \
+		uniform vec4 albedo : hint_color = vec4(1,1,1,1); \
+		uniform sampler2D texture_albedo : hint_albedo; \
+		void vertex() { \
+			if (!OUTPUT_IS_SRGB) { \
+				COLOR.rgb = mix( pow((COLOR.rgb + vec3(0.055)) * (1.0 / (1.0 + 0.055)), vec3(2.4)), COLOR.rgb* (1.0 / 12.92), lessThan(COLOR.rgb,vec3(0.04045)) ); \
+			} \
+			VERTEX = VERTEX; \
+			POSITION=PROJECTION_MATRIX*INV_CAMERA_MATRIX*WORLD_MATRIX*vec4(VERTEX.xyz,1.0); \
+			POSITION.z = mix(POSITION.z, 0, 0.998); \
+		} \
+		void fragment() { \
+			vec2 base_uv = UV; \
+			vec4 albedo_tex = texture(texture_albedo,base_uv); \
+			albedo_tex *= COLOR; \
+			ALBEDO = albedo.rgb * albedo_tex.rgb; \
+			if (albedo.a * albedo_tex.a < 0.5) { discard; } \
+			ALPHA = albedo.a * albedo_tex.a; \
+		} \
+	");
```
This whole shader should be simplified, the `vertex()` code is correct, but we don't need any of the uniforms and the fragment color just needs to set the ALBEDO. Something like this is probably enough: 

```
void fragment() {
    ALBEDO = COLOR;
}
```

TokageItLab replied:
Related to the hack of the drawing order by Vertex, it is necessary to put it on Alpha's rendering pipeline to prevent Gizmo from casting shadows on the objects. So, finally
```
void fragment() { \
	ALBEDO = COLOR.rgb; \
	ALPHA = COLOR.a; \
} \
```

JFonS commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -666,22 +804,255 @@ void Skeleton3DEditor::_bind_methods() {
+	handle_material = Ref<ShaderMaterial>(memnew(ShaderMaterial));
+	handle_shader = Ref<Shader>(memnew(Shader));
+	handle_shader->set_code(" \
+		shader_type spatial; \
+		render_mode unshaded, shadows_disabled, depth_draw_always; \
+		uniform vec4 albedo : hint_color = vec4(1,1,1,1); \
+		uniform sampler2D texture_albedo : hint_albedo; \
+		uniform float point_size : hint_range(0,128) = 32; \
+		void vertex() { \
+			if (!OUTPUT_IS_SRGB) { \
+				COLOR.rgb = mix( pow((COLOR.rgb + vec3(0.055)) * (1.0 / (1.0 + 0.055)), vec3(2.4)), COLOR.rgb* (1.0 / 12.92), lessThan(COLOR.rgb,vec3(0.04045)) ); \
+			} \
+			VERTEX = VERTEX; \
+			POSITION=PROJECTION_MATRIX*INV_CAMERA_MATRIX*WORLD_MATRIX*vec4(VERTEX.xyz,1.0); \
+			POSITION.z = mix(POSITION.z, 0, 0.999); \
+			POINT_SIZE = point_size; \
+		} \
+		void fragment() { \
+			vec4 albedo_tex = texture(texture_albedo,POINT_COORD); \
+			vec3 col = albedo_tex.rgb + COLOR.rgb; \
+			col = vec3(min(col.r,1.0),min(col.g,1.0),min(col.b,1.0)); \
+			ALBEDO = albedo.rgb * col; \
+			if (albedo.a * albedo_tex.a < 0.5) { discard; } \
+			ALPHA = albedo.a * albedo_tex.a; \
+		} \
```
The `albedo` uniform is unused, it should be removed. The rest looks great. 

There are plans to add an easier way to do this kind of Z-scaling, but until we add it, this is the only way to achieve the wanted effect.

JFonS reviewed (commented):
Reviewed. Overall looks good, but I left some comments.

Other than that, I noticed that the plugin only allows editing one bone at a time. This is fine, but it creates some weird behavior when you have one bone selected and "shift-click" another one to add it to the selection. One of them is drawn in red as selected, but the transform handle moves to the other one. We can look into adding an overrideable method (similar to `is_selectable_when_hidden()`) that tells the editor a gizmo doesn't support multiple subgizmos. This isn't critical, so it can be done in a separate PR, just writing it down so I don't forget about it :)

TokageItLab commented on editor/editor_plugin.h:
```diff
@@ -176,6 +176,12 @@ class EditorPlugin : public Node {
+		DISCARD,
+		UNSELECT
+	};
+
```
For now, I renamed it to `AFTER_GUI_INPUT_PASS`, `AFTER_GUI_INPUT_STOP`, `AFTER_GUI_INPUT_DESELECT`.

TokageItLab commented on doc/classes/EditorPlugin.xml:
```diff
@@ -178,22 +178,22 @@
 				public override bool ForwardSpatialGuiInput(Camera3D camera, InputEvent @event)
 				{
-				    return true;
+				    return EditorPlugin.DISCARD;
```
It has been changed on my side, but maybe the push was failed... For now, make the changes including renaming the Enum as you reviewed below.

TokageItLab commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -692,16 +1063,374 @@ void EditorInspectorPluginSkeleton::parse_begin(Object *p_object) {
+	ERR_FAIL_COND(!skeleton);
+
+	if (!Skeleton3DEditor::get_singleton()->is_edit_mode() || skeleton->get_selected_bone() == -1) {
+		skeleton->set_transform(p_transform);
```
There were two other parts that did the same decision, so I removed them.

TokageItLab replied:
This change causes some weird behavior, so it may be necessary to look into SubGizmo.

TokageItLab replied:
@JFonS I noticed that I need to do `clear_subgizmo_selection()` when the return value of `subgizmos_intersect_ray()` is `-1`. Do you think it's a bad way to do that upstream (in `Node3DEditorPlugin`)?

TokageItLab commented:
@JFonS I fixed the reviewed parts and re-based it with updates by SkeletonIK. There are no particular conflicts with SkeletonIK. However, if SkeletonGizmo can be used in combination with SkeletonIK, it will be a more powerful feature, and I expect it to be :) But, that is outside the scope of this PR.

CC @fire @TwistedTwigleg

TokageItLab commented:
And I'll work the "shift-click" issue later.

TokageItLab commented:
@TwistedTwigleg Excuse me, the `Array process_order` was removed from `Skeleton` by the SkeletonIK change, so does that mean that the internal idx and the visual idx are no longer mismatched?

TokageItLab commented:
It's a bit complicated, so I revert to pre-merge.

TokageItLab commented:
As mentioned in https://github.com/godotengine/godot/pull/51368#issuecomment-900738166, the rendering of the octahedron in SkeletonGizmo's bones is broken, so I'm fixing that ahead.

TokageItLab commented:
@JFonS CC @fire 

1. Changed SkeletonGizmo's drawing algorithm so that it no longer uses skinning, thus skew no longer occurs and fixed vulkan error when selecting skeleton.

2. Since https://github.com/godotengine/godot/pull/51459#issuecomment-903298629 is not merged, the insert animation function is still broken, but I can fix the other one later after one of them is merged.

3. For the previous review comment of https://github.com/godotengine/godot/pull/45699#discussion_r689606750, I purposely changed it to display an error, which looks like a SubGizmo error but is not a fatal error and can be fixed later.

```cpp
Transform3D Skeleton3DGizmoPlugin::get_subgizmo_transform(const EditorNode3DGizmo *p_gizmo, int p_id) const {
    Skeleton3D *skeleton = Object::cast_to<Skeleton3D>(p_gizmo->get_spatial_node());
    ERR_FAIL_COND_V(!skeleton, Transform3D());

    // Probably a bug: get_subgizmo_transform() fires
    // even if you call clear_subgizmo_selection() in subgizmos_intersect_ray().
    ERR_FAIL_COND_V(skeleton->get_selected_bone() == -1, Transform3D());

    return skeleton->get_bone_global_pose(skeleton->get_selected_bone());
}
```

TokageItLab commented:
@JFonS CC @fire

#51722 has been merged, and pose and custom posed editor have been fixed to be read only correctly. There are some issues with "Other Issues (Todo in the future)" in the description and [latest comments](https://github.com/godotengine/godot/pull/45699#issuecomment-908778701), but they can be fixed later, and I think this PR is ready to be merged.

reduz commented:
Is it a good idea to draw the skeleton not using skinning? I have the feeling this may have quite serious performance implications

TokageItLab commented:
@reduz You have a point. However, it is never good to have skew in Gizmo as a result of using skinning.

Currently, gizmo uses ArrayMesh, so if we use ImmediateMesh, we can expect some performance improvement, but we will probably need to change not only skeleton but all child class of Node3D gizmo at the same time.

Also, the SkeletonGizmo is only updated when tools is true, so there is no need to worry about performance in general games.

TokageItLab commented:
If we really need to use skinning, we should use a Gizmo instead of an octahedron, which does not cause skew, and Blender's Gizmo may be a good reference.

reduz commented:
I agree we should change the gizmo to avoid the skew, I suppose any shape that works makes sense.

TokageItLab commented:
@reduz When using skinning, the only shape that doesn't cause Skew is the Wire shape, anyway, but how do you feel about adopting that?

<img width="643" alt="スクリーンショット 2021-09-07 0 44 10" src="https://user-images.githubusercontent.com/61938263/132240722-ad2b480f-f577-4988-b2e0-d34b1f5bef5a.png">

TokageItLab commented:
Also, there is an error that vulkan's free() fails due to skinning. As I mentioned above, there is no performance concern except for tools, so I think one way is to use a method that does not use skinning.

TokageItLab commented:
There are 4 possible ways to prevent Skew. 

1. As it is now, we omit Skinning and redraw gizmo when pose_update.

2. Use Skinning and adopt the Wire shape.

3. Use Skinning to define an arbitrary shape as one of the `RenderingServer::PrimitiveType` and draw it on the `RenderingServer`.

4. Implement Compute Shaders and draw arbitrary shapes with Shaders.

@reduz 3 and 4 are out of the scope of this PR. If you plan to adopt these, it makes sense to use 2. However, in terms of ease of viewing and using, I would recommend 1.

lyuma commented:
I would like to note that the heavy work in `Skeleton3DGizmoPlugin::redraw` is the call to `get_bone_global_pose` since this talks to the RenderingServer and the `.inverse()` on each basis for drawing the coordinate axes.

If there is performance concern, I'd suggest just adding a `LocalVector<Transform3D>` at the top for calculating the bone global poses directly to avoid the expensive `get_global_bone_pose`, and maybe only do the inverse() once per bone, rather than once per axis.

If this is done, it's basically a small and constant amount of vector math per bone, the heaviest of which is 6 cross products: 2 per axis. (As well as the cost of calculating global bone pose matrices, which is unavoidable.)

In terms of the number of vertices, it looks something like 38 per vertex. Assuming a rig with 300 bones, that's 10000 vertices produced per frame, which should be no problem on the CPU. It's also very much proportional to the cost of calculating the 300 bone matrices on the CPU that would need to be copied to the GPU for a skinning pass.

Therefore, the way things are, I do not think a skinning pass is a worthwhile option to pursue. It will become necessary when we have scenes with hundreds of skeletons in a single editor scene. I'd like to wait for the usecase to be presented so we have a real world benchmark to base performance improvements on.

TokageItLab commented:
As Lyuma said, it makes sense to do it in a way that avoids access to `get_bone_global_pose()`.

However, there have been PRs posted in the past with suggestions for improving `get_bone_global_pose()`, but which is currently in limbo.

https://github.com/godotengine/godot/pull/41322
https://github.com/godotengine/godot/pull/42681

I think it is necessary to salvage them anyway, and I think current methods will be possible to adapt to them salvaged later, and which is good for other methods that use `get_bone_global_pose()`.

reduz commented:
@TokageItLab If I have to be completely honest, and thinking about it, I dont mind the bone skewing very much, I think this pretty much information anyway, and its rare to actually see this happening in practice. If you do, it may also be an indicative of a mistake or error in the model. Godot is not really a 3D modelling app anyway.

@lyuma it does not scale, its common to see complex models use thousands of bones for stuff like cinematics, we need to entirely avoid doing this using CPU.

TokageItLab commented:
@reduz "Currently", you may not notice the skewed gizmo, because the gizmo is not shown in the foreground, and it is buried in the model. Also, it is not easy to manipulate the bones.

With the implementation of the Gizmo Manipulator, the selected SkeletonGizmo is now in the foreground, and making it easier to manipulate the bones. It may makes sense for skew to occur with Scaling, but I think it's completely wrong for skew to occur with Translating. Also, although Godot is not a 3D modeling software, it is necessary to understand the bone structure correctly in order to animate it.

TokageItLab commented:
@reduz At least https://github.com/godotengine/godot/pull/51368#issuecomment-900738166 has changed the behavior of skinning, so gizmo transformation using skinning is broken as shown in the link comment video. It is possible to revert to broken behavior, but it is necessary to look into where the skinning was broken in order to correctly transform the gizmo using skinning.

TokageItLab commented:
I've reverted to gizmo with skinning once, although the skew problem recurs. It is may possible to improve it later

TokageItLab commented:
It seems that reduz is thinking of getting rid of the bone rest. It may be not good to allow users to easily handle posing with bone rests. However, when it comes to retargeting or adjusting axes, rest needs to be changed.

Therefore, it would be better to omit the implementation of rest mode bone manipulation, and make rest and custom poses read only in the inspector, and instead add a "Rest Option" in the 3D View menu, such as operating "Unify roll axis to one axis", "Swap axis" and etc.

lyuma commented:
As a followup to the idea of removing bone_rest as it currently is... and just jotting down some thoughts I had recently: I've been increasingly thinking that rest_pose should become an animation concept of some sort, and added to "rotation", "position" and "scale" type tracks.

Basically, Animations could use the rest matrix (transform/rotation/scale) applied one after the other, so there is no problem with non-uniform matrices but you can still assign "identity" tracks in the animation to reset the object to the rest pose.

Another option is for the animation transform track to be separated into animation position track, rotation track and scale track (note that a Transform track internally already stores data in this form, but it requires all 10 values to be set, even though most animations only change rotation)... that way, the position would remain unchanged and the bone rest does not need to be stored.

TokageItLab commented:
Updated:
- Omit rest mode
- Rest and Custom Pose are forced readonly in inspector
- Bone shape is able to be changed in EditorSetting
- Implemented ShowRestOnly mode in Skeleton

JFonS reviewed (changes requested):
Looks good in general, amazing work :)

I left some comments on minor details and some changes that need to be made, but the overall design is good.

I also found a crash while testing. It can be reproduced with an animated skeleton:
1. Have a scene with a skeleton and an AnimationPlayer that contains an animation for the skeleton.
2. Open animation editor with and the skeleton animation.
3. Select the animated skeleton, so the gizmo is drawn (the animation doesn't need to be playing).
4. Close the scene.
5. Crash happens.

I fixed it with the following changes:
```
diff --git a/editor/plugins/skeleton_3d_editor_plugin.cpp b/editor/plugins/skeleton_3d_editor_plugin.cpp
--- a/editor/plugins/skeleton_3d_editor_plugin.cpp	
+++ b/editor/plugins/skeleton_3d_editor_plugin.cpp	
@@ -356,6 +356,7 @@
 
 void Skeleton3DEditor::_update_show_rest_only() {
 	_update_pose_enabled(-1);
+	_update_gizmo_visible();
 }
 
 void Skeleton3DEditor::_update_pose_enabled(int p_bone) {
@@ -370,7 +371,6 @@
 			pose_editor->set_transform_read_only(skeleton->is_show_rest_only() || !(skeleton->is_bone_enabled(selected)));
 		}
 	}
-	_update_gizmo_visible();
 }
 
 void Skeleton3DEditor::_on_click_skeleton_option(int p_skeleton_option) {
@@ -667,6 +667,7 @@
 	set_rest_options_enabled(selected);
 	_update_properties();
 	_update_pose_enabled();
+	_update_gizmo_visible();
 }
 
 // May be not used with single select mode.
```

But please check if the fix makes sense and feel free to change it if it doesn't.

TokageItLab commented on editor/plugins/node_3d_editor_plugin.h:
```diff
@@ -756,7 +758,8 @@ class Node3DEditor : public VBoxContainer {
 
 	Transform3D get_gizmo_transform() const { return gizmo.transform; }
-	bool is_gizmo_visible() const { return gizmo.visible; }
+	void set_gizmo_visible(bool p_visible) { gizmo.visible = p_visible; }
+	bool is_gizmo_visible() const { return gizmo.enable && gizmo.visible; }
```
I agree and I will work on it.

TokageItLab commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -671,25 +870,202 @@ void Skeleton3DEditor::_bind_methods() {
+	pointsm->hide();
+}
+
+void Skeleton3DEditor::_draw_gizmo() {
```
This is like a wrapper function to avoid errors due to bone update timing. It might be better to rename the function.

TokageItLab replied:
For now, I'll leave it in place.

TokageItLab commented on scene/3d/skeleton_3d.cpp:
```diff
@@ -1140,6 +1160,16 @@ void Skeleton3D::execute_modifications(real_t p_delta, int p_execution_mode) {
 
 #endif // _3D_DISABLED
 
+void Skeleton3D::set_selected_bone(int p_bone) {
+	selected_bone = p_bone;
+	update_gizmos();
+	return;
+}
+
+int Skeleton3D::get_selected_bone() const {
+	return selected_bone;
+}
+
```
This is a remnant of the time when the Subgizmo system was not yet implemented. Now there is certainly no need to store it in the skeleton. I fix it.

TokageItLab commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -700,16 +1076,411 @@ void EditorInspectorPluginSkeleton::parse_begin(Object *p_object) {
+	// even if you call clear_subgizmo_selection() in subgizmos_intersect_ray().
+	ERR_FAIL_COND_V(skeleton->get_selected_bone() == -1, Transform3D());
+
+	return skeleton->get_bone_global_pose(skeleton->get_selected_bone());
```
I remember doing it this way because it was difficult to create unselected states. I will try to see if I can do something about it.

TokageItLab replied:
There was a compatibility problem with the Bones tree in the inspector, and I solved it.

TokageItLab commented:
@JFonS Thank you for your reviewing and I fixed the part you reviewed.

Also, as for the crashing bug, it seems that the problem was caused by the editor generation during scene transition. That method seemed unnecessary, so I've solved it by omitting it.

I'll squash it if it seems to be OK to you.

JFonS reviewed (approved):
Re-reviewed and all my concerns were addressed.

Thanks for your great work and for putting up with my delayed reviews :)

Shadowblitz16 commented:
Why does this use embeded bone data instead of BoneAttachment3D?

TokageItLab commented:
I do not understand what you are asking. FYI, as to why not all bones should be children of the scene. It is written in https://github.com/godotengine/godot-proposals/issues/1767 because it is mainly from a performance point of view.

If you want to generate BoneAttachment from multiple bones, I remember it is possible to add it when setting up PhysicalBone, but I don't think there was a way to add only BoneAttachment. You can open a proposal for that.

If you want to convert a gizmo to a MeshInstance, that doesn't make sense. Most 3D applications don't have that capability and Godot is the same. That doesn't have to be included in the core as a general feature. I know that Blender has such a thing as an 3rd party addon, so I suggest you look for an addon or make your own.
